src/Compatibility/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs

```diff
@@ -16,7 +15,7 @@ public class Suppression : IEquatable<Suppression>
         /// <summary>
         /// The DiagnosticId representing the error to be suppressed.
         /// </summary>
-        public string? DiagnosticId { get; set; }
+        public string DiagnosticId { get; set; }
```

Comment:
I may be miss-remembering, but I believe the original reason why DiagnosticId was nullable is because we wanted to be able to have global suppressions in the future where you could suppress all errors for a particular pair of left and right, without the need to call out each individual DiagnosticId being suppressed.

Reply:
I nearly pulled my head out to understand how this could ever possible :D. From what I can see, a DiagnosticId is always required, in any case. A suppression created by PackageValidation, ApiCompatibility or any future component must always contain a DiagnosticId.

The suppression is passed to the [`IsErrorSuppressed` method](https://github.com/dotnet/sdk/blob/c8cd2cfbbe7e2977d20177c03472add5871c4c1a/src/Compatibility/Microsoft.DotNet.ApiCompatibility/ErrorSuppression/SuppressionEngine.cs#L62). From what I noticed, the `error.DiagnosticId == null` check was dead code.

Based on the above, a global suppression is currently defined either as:
- DiagnosticId + Target + IsBaselineSuppression
- DiagnosticId + Left + Right + IsBaselineSuppression

Based on that, a global suppression is either a suppression of a specific target applying to all lefts and rights or a suppression of all targets applying to a specific left and right.

@safern hope it's okay to ask you if this matches your memory?

Reply:
If having a global suppression (without diagnostic ID) wasn't supported before, that was not by design, and was likely a bug. As you can see on the discussion thread for this issue: https://github.com/dotnet/sdk/issues/22901 Having the ability to do global suppression for a left + right pair is something that people will want to have, so I would still push on the fact of reverting your change and have DiagnosticId be nullable again.

Reply:
My point is that we want to make sure that diagnostic id is always provided by compatibility tooling. I agree that a customer might want to globally suppress a left and a right (even though that raises interesting questions like why even compare something that is already suppressed).

I submitted https://github.com/dotnet/sdk/pull/25938 that updates the suppression engine to support this new type of global suppression. Support for what you are asking is added while still maintaining the desire to make sure that components need to provide a suppression id (so that suppressions are uniquely identifiable and suppressable => i.e. via NoWarn property).